File: tests/widgets/controls/combobox.nim
import std/[os, strutils]

import knutella/appkit
import knutella/objc

proc maxFramesFromEnv(defaultValue = -1): int =
  let raw = getEnv("KNUTELLA_EXAMPLE_FRAMES").strip()
  if raw.len == 0:
    return defaultValue
  try:
    parseInt(raw)
  except ValueError:
    defaultValue

proc debugRenderDumpEnabled(): bool =
  getEnv("KNUTELLA_APPKIT_DEBUG_RENDER").strip().toLowerAscii() in
    ["1", "true", "yes", "on"]

proc format1(value: float32): string =
  formatFloat(value.float, ffDecimal, 1)

proc formatRect(rect: NSRect): string =
  "(" & format1(rect.origin.x) & "," & format1(rect.origin.y) & " " &
    format1(rect.size.width) & "x" & format1(rect.size.height) & ")"

proc intBool(flag: bool): int =
  if flag: 1 else: 0

proc dumpViewLine(view: NSView, name: string) =
  if view.isNil:
    return
  echo "[",
    name,
    "] frame=",
    formatRect(view.frame()),
    " bounds=",
    formatRect(view.bounds()),
    " autoresizeMask=0x",
    toHex(view.autoresizingMask(), 2)

proc dumpComboBoxLine(comboBox: NSComboBox, name: string) =
  if comboBox.isNil:
    return
  dumpViewLine(comboBox.NSView, name)
  echo "[",
    name,
    "] editable=",
    intBool(comboBox.isEditable()),
    " selected=",
    comboBox.indexOfSelectedItem(),
    " numberOfItems=",
    comboBox.numberOfItems(),
    " string='",
    comboBox.stringValue(),
    "'"

proc dumpComboBoxLayout(
    stage: string, window: NSWindow, comboBox1: NSComboBox, comboBox2: NSComboBox
) =
  if window.isNil:
    return
  let windowFrame = window.frame()
  let contentRect = window.contentRectForFrameRect(windowFrame)
  let contentView = window.contentView()
  echo "[Window ",
    stage, "] frame=", formatRect(windowFrame), " contentRect=", formatRect(contentRect)
  dumpViewLine(contentView, "contentView")
  dumpComboBoxLine(comboBox1, "comboBox1")
  dumpComboBoxLine(comboBox2, "comboBox2")
  if debugRenderDumpEnabled():
    debugDumpWindowRenderTree(window)

objcImpl:
  type ComboBoxWindow = object of NSWindow
    xComboBox1: NSComboBox
    xComboBox2: NSComboBox

  method init*(self: var ComboBoxWindow): ComboBoxWindow =
    var base = self.initWithContentRect(
      100.0,
      100.0,
      300.0,
      328.0,
      {NSTitledWindow, NSClosableWindow, NSMiniaturizableWindow, NSResizableWindow},
      NSBackingStoreBuffered,
      false,
    )
    result = asTypeRaw[ComboBoxWindow](base.value)
    base.value = nil
    if result.isNil:
      return

    var contentAlloc = NSView.alloc()
    var contentView = contentAlloc.initWithFrame(nsRect(0.0, 0.0, 300.0, 300.0))
    contentAlloc.value = nil
    result.setContentView(contentView)

    var comboBox1Alloc = NSComboBox.alloc()
    result.xComboBox1 = comboBox1Alloc.init()
    comboBox1Alloc.value = nil
    result.xComboBox1.setFrame(nsRect(10.0, 260.0, 121.0, 26.0))
    result.xComboBox1.addItemWithObjectValue(@ns"item1")
    result.xComboBox1.addItemWithObjectValue(@ns"item2")
    result.xComboBox1.addItemWithObjectValue(@ns"item3")
    result.xComboBox1.setTarget(ID(value: result.value))
    result.xComboBox1.setAction(getSelector("OnComboBox1SelectedItemChange:"))
    result.xComboBox1.selectItemAtIndex(1)

    var comboBox2Alloc = NSComboBox.alloc()
    result.xComboBox2 = comboBox2Alloc.init()
    comboBox2Alloc.value = nil
    result.xComboBox2.setFrame(nsRect(10.0, 220.0, 121.0, 26.0))
    result.xComboBox2.setEditable(false)
    result.xComboBox2.addItemWithObjectValue(@ns"item1")
    result.xComboBox2.addItemWithObjectValue(@ns"item2")
    result.xComboBox2.addItemWithObjectValue(@ns"item3")
    result.xComboBox2.setTarget(ID(value: result.value))
    result.xComboBox2.selectItemAtIndex(1)

    contentView.addSubview(result.xComboBox1)
    contentView.addSubview(result.xComboBox2)
    result.setTitle(@ns"ComboBox Example")
    result.setIsVisible(true)
    dumpComboBoxLayout("init", result as NSWindow, result.xComboBox1, result.xComboBox2)
    contentView.value = nil

  method windowShouldClose*(self: ComboBoxWindow, sender: NSObject): bool =
    if sender.isNil:
      NSApp().stop()
      return true
    NSApp().stop()
    true

  method OnComboBox1SelectedItemChange*(self: ComboBoxWindow, sender: NSObject) =
    if sender.isNil:
      return
    self.xComboBox2.selectItemAtIndex(self.xComboBox1.indexOfSelectedItem())
    dumpComboBoxLayout(
      "comboBox1-change", self as NSWindow, self.xComboBox1, self.xComboBox2
    )

  method OnComboBox2SelectedItemChange*(self: ComboBoxWindow, sender: NSObject) =
    if sender.isNil:
      return
    self.xComboBox1.selectItemAtIndex(self.xComboBox2.indexOfSelectedItem())
    dumpComboBoxLayout(
      "comboBox2-change", self as NSWindow, self.xComboBox1, self.xComboBox2
    )

  method dealloc(self: ComboBoxWindow) {.used.} =
    self.xComboBox1 = NSComboBox(value: nil)
    self.xComboBox2 = NSComboBox(value: nil)
    destroyIvarFields(self)
    discard callSuperIdFrom(ComboBoxWindow, self, getSelector("dealloc"))

when isMainModule:
  var app = NSApp()
  var windowAlloc = ComboBoxWindow.alloc()
  var window = initOwned(move(windowAlloc))

  app.addWindow(window.NSWindow)
  window.makeKeyAndOrderFront(app.NSObject)

  if getEnv("KNUTELLA_COMBOBOX_OPEN_POPUP").strip().len > 0:
    window.xComboBox1.openPopup()
    dumpComboBoxLayout(
      "popup-open", window.NSWindow, window.xComboBox1, window.xComboBox2
    )

  if getEnv("KNUTELLA_COMBOBOX_TRIGGER_CALLBACK").strip().len > 0:
    dumpComboBoxLayout(
      "post-front", window.NSWindow, window.xComboBox1, window.xComboBox2
    )
    window.OnComboBox1SelectedItemChange(window.NSObject)

  try:
    let maxFrames = maxFramesFromEnv()
    if maxFrames < 0:
      app.run()
    else:
      discard app.runForFrames(maxFrames)
  except CatchableError as exc:
    echo "Unable to run combobox example: ", exc.msg

  window.value = nil
  app.value = nil

File: examples/widgets/controls/ComboBox.m
#include <Cocoa/Cocoa.h>
#include <stdlib.h>

@interface Window : NSWindow  {
  NSComboBox* comboBox1;
  NSComboBox* comboBox2;
}
- (instancetype)init;
- (void)dumpLayout:(NSString *)stage;
- (void)dumpAndTriggerForPorting;
- (BOOL)windowShouldClose:(id)sender;
@end

@implementation Window
- (int)intBool:(BOOL)flag {
  return flag ? 1 : 0;
}

- (void)dumpComboBox:(NSComboBox *)comboBox name:(const char *)name {
  NSRect frame = [comboBox frame];
  NSRect bounds = [comboBox bounds];
  NSLog(@"[%s] frame=(%.1f,%.1f %.1fx%.1f) bounds=(%.1f,%.1f %.1fx%.1f) autoresizeMask=0x%02lx",
    name,
    frame.origin.x, frame.origin.y, frame.size.width, frame.size.height,
    bounds.origin.x, bounds.origin.y, bounds.size.width, bounds.size.height,
    (unsigned long)[comboBox autoresizingMask]);
  NSLog(@"[%s] editable=%d selected=%ld numberOfItems=%ld string='%@'",
    name,
    [self intBool:[comboBox isEditable]],
    (long)[comboBox indexOfSelectedItem],
    (long)[comboBox numberOfItems],
    [comboBox stringValue]);
}

- (void)dumpLayout:(NSString *)stage {
  NSRect frame = [self frame];
  NSRect contentRect = [self contentRectForFrameRect:frame];
  NSView *content = [self contentView];
  NSRect contentFrame = [content frame];
  NSRect contentBounds = [content bounds];
  NSLog(@"[Window %@] frame=(%.1f,%.1f %.1fx%.1f) contentRect=(%.1f,%.1f %.1fx%.1f)",
    stage,
    frame.origin.x, frame.origin.y, frame.size.width, frame.size.height,
    contentRect.origin.x, contentRect.origin.y, contentRect.size.width, contentRect.size.height);
  NSLog(@"[contentView] frame=(%.1f,%.1f %.1fx%.1f) bounds=(%.1f,%.1f %.1fx%.1f) autoresizeMask=0x%02lx",
    contentFrame.origin.x, contentFrame.origin.y, contentFrame.size.width, contentFrame.size.height,
    contentBounds.origin.x, contentBounds.origin.y, contentBounds.size.width, contentBounds.size.height,
    (unsigned long)[content autoresizingMask]);
  [self dumpComboBox:comboBox1 name:"comboBox1"];
  [self dumpComboBox:comboBox2 name:"comboBox2"];
}

- (instancetype)init {
  self = [super initWithContentRect:NSMakeRect(100, 100, 300, 300) styleMask:NSWindowStyleMaskTitled | NSWindowStyleMaskClosable | NSWindowStyleMaskMiniaturizable | NSWindowStyleMaskResizable backing:NSBackingStoreBuffered defer:NO];
  if (self == nil) {
    return nil;
  }

  comboBox1 = [[[NSComboBox alloc] initWithFrame:NSMakeRect(10, 260, 121, 26)] autorelease];
  [comboBox1 addItemWithObjectValue:@"item1"];
  [comboBox1 addItemWithObjectValue:@"item2"];
  [comboBox1 addItemWithObjectValue:@"item3"];
  [comboBox1 setTarget:self];
  [comboBox1 setAction:@selector(OnComboBox1SelectedItemChange:)];
  //[[comboBox1 delegate] comboBoxSelectionIsChanging: ];
  //[comboBox1 setDelegate:self];
  [comboBox1 selectItemAtIndex:1];
  
  comboBox2 = [[[NSComboBox alloc] initWithFrame:NSMakeRect(10, 220, 121, 26)] autorelease];
  [comboBox2 setEditable:false];
  [comboBox2 addItemWithObjectValue:@"item1"];
  [comboBox2 addItemWithObjectValue:@"item2"];
  [comboBox2 addItemWithObjectValue:@"item3"];
  [comboBox2 setTarget:self];
  //[comboBox2 setAction:@selector(OnComboBox2SelectedItemChange)];
  [comboBox2 selectItemAtIndex:1];

  [self setTitle:@"ComboBox Example"];
  [[self contentView] addSubview:comboBox1];
  [[self contentView] addSubview:comboBox2];
  [self setIsVisible:YES];
  [self dumpLayout:@"init"];
  return self;
}

- (BOOL)windowShouldClose:(id)sender {
  [NSApp terminate:sender];
  return YES;
}

- (IBAction) OnComboBox1SelectedItemChange:(id)sender {
  [comboBox2 selectItemAtIndex:[comboBox1 indexOfSelectedItem]];
  [self dumpLayout:@"comboBox1-change"];
}

- (IBAction) OnComboBox2SelectedItemChange:(id)sender {
  [comboBox1 selectItemAtIndex:[comboBox2 indexOfSelectedItem]];
  [self dumpLayout:@"comboBox2-change"];
}

- (void)dumpAndTriggerForPorting {
  [self dumpLayout:@"post-front"];
  [self OnComboBox1SelectedItemChange:self];
  [NSApp terminate:self];
}

- (void)OnComboBox1SelectionIsChanging:(NSNotification *)notification {
  
}
- (void)comboBoxSelectionIsChanging:(NSNotification *)notification {
  
}

@end

int main(int argc, char *argv[]) {
    [NSApplication sharedApplication];
    [NSApp setActivationPolicy:NSApplicationActivationPolicyRegular];

    Window *window = [[Window alloc] init];
    [window makeKeyAndOrderFront:nil];

    if (getenv("NUTELLA_COMBOBOX_EXIT_AFTER_DUMP")) {
      [NSApp activateIgnoringOtherApps:YES];
      [window performSelector:@selector(dumpAndTriggerForPorting) withObject:nil afterDelay:0.05];
      [NSApp run];
      return 0;
    }

    [NSApp activateIgnoringOtherApps:YES];
    [NSApp run];
    return 0;
}
